tools/applets/schedule.py

```diff
@@ -0,0 +1,210 @@
+
+        cw = QtGui.QFontMetrics(self.font()).averageCharWidth()
+        h = self.table.horizontalHeader()
+        h.resizeSection(0, 7 * cw)
+        h.resizeSection(1, 12 * cw)
+        h.resizeSection(2, 16 * cw)
+        h.resizeSection(3, 6 * cw)
+        h.resizeSection(4, 16 * cw)
+        h.resizeSection(5, 30 * cw)
+        h.resizeSection(6, 20 * cw)
+        h.resizeSection(7, 20 * cw)
+
+    def _create_context_menu(self):
+        self.table.setContextMenuPolicy(QtCore.Qt.ActionsContextMenu)
+        request_termination_action = QtWidgets.QAction(
+            "Request termination", self.table
+        )
+        request_termination_action.triggered.connect(partial(self.delete_clicked, True))
+        request_termination_action.setShortcut("DELETE")
+        request_termination_action.setShortcutContext(QtCore.Qt.WidgetShortcut)
+        self.table.addAction(request_termination_action)
+
+        delete_action = QtWidgets.QAction("Delete", self.table)
+        delete_action.triggered.connect(partial(self.delete_clicked, False))
+        delete_action.setShortcut("SHIFT+DELETE")
+        delete_action.setShortcutContext(QtCore.Qt.WidgetShortcut)
+        self.table.addAction(delete_action)
+
+        terminate_pipeline = QtWidgets.QAction(
+            "Gracefully terminate all in pipeline", self.table
+        )
+        terminate_pipeline.triggered.connect(self.terminate_pipeline_clicked)
+        self.table.addAction(terminate_pipeline)
+
+        initialization = QtWidgets.QAction("Run initalization experiment", self.table)
+        initialization.triggered.connect(self.run_initialization)
+        self.table.addAction(initialization)
+
+        background = QtWidgets.QAction("Run background experiment", self.table)
+        background.triggered.connect(self.run_background)
+        self.table.addAction(background)
+
+    async def connect_subscribers(self):
+        localhost = "::1"
+        port_notify = 3250
+        self._schedule_sub = ModelSubscriber(
+            "schedule", schedule.Model, self._report_disconnect
+        )
+        await self._schedule_sub.connect(localhost, port_notify)
+
+    def _report_disconnect(self):
+        if not self._disconnected_reported:
+            logging.error("connection to master lost, restart dashboard to reconnect")
+            self.setDisabled(True)
+        self._disconnect_reported = True
+
+    async def labrad_connected(self):
+        await self.artiq_connected()
+        await self.setup_cxn_listeners()
+
+    async def setup_cxn_listeners(self):
+        self.cxn.add_on_connect("artiq", self.run_in_labrad_loop(self.artiq_connected))
+        self.cxn.add_on_disconnect("artiq", self.artiq_disconnected)
+
+    async def artiq_connected(self):
+        self.artiq = self.cxn.get_server("artiq")
+        if not self._disconnect_reported:
+            self.setDisabled(False)
+
+    def artiq_disconnected(self):
+        self.setDisabled(True)
```

Comment:
should this also call self._report_disconnect()

Reply:
I don't think this function should call report_disconnected. The function is called when the artiq server disconnects. report_disconnected is for when the artiq subscribers disconnect.